packages/kotti-ui/source/props.ts

```diff
@@ -0,0 +1,198 @@
+	['ZodNumber', Number],
+	['ZodObject', Object],
+	['ZodString', String],
+])
+
+/**
+ * @summary Easily define & validate vue props, improve error messages drastically
+ *
+ * @description
+ * This augments/replaces Vue’s props entirely, including the need to add prop.validator
+ * by using zod and therefore actually prints errors to the console that can help developers
+ * rather than just say "hey you failed the validator, good luck!" as vue would do by default.
+ *
+ * ## Known limitations:
+ *
+ * 1. Deeply defined defaults are only applied when manually `schema.safeParse`-ing the data (only top-level defaults are possible to transform to Vue)
+ * 2. While this function also automatically types the return types via as PropType<> (which in theory
+ *   should mean that annotating the prop types explicitly is no longer necessary) in `@vue/composition-api`
+ *   (`0.6.1` and `1.1.5`), `PropType<>` appears to be bugged.
+ *   For example KtBreadcrumb’s props get inferred to be any
+ *   So, unfortunately, this may only help in the future, once the Vue team fixes this
+ * 3. `z.function().default()` does not accept `undefined` while parsing the schema manually
+ *   without an explicit `.optional()` {@link https://github.com/colinhacks/zod/issues/647}
+ *   However, it can still be used to generate the Vue defaults, as Vue doesn’t call validators with `undefined`
+ *
+ * @example
+ * // KtUserMenu.vue
+ * export default defineComponent<KottiUserMenu.PropsInternal>({
+ *   name: 'KtUserMenu',
+ *   props: makeProps(KottiUserMenu.propsSchema)
+ * })
+ *
+ * // types.ts
+ * export namespace KottiUserMenu {
+ *   const propsSchema = z.object({
+ *     // ...
+ *   })
+ *   type Props = z.input<typeof propsSchema>
+ *   type PropsInternal = z.output<typeof propsSchema>
+ * }
+ */
+export const makeProps = <SCHEMA extends z.ZodObject<z.ZodRawShape>>(
+	schema: SCHEMA,
+): {
+	[KEY in keyof SCHEMA['shape']]: Omit<PropOptions, 'required' | 'type'> & {
+		required: z.output<SCHEMA>[KEY] extends undefined ? false : true
+		type: PropType<z.output<SCHEMA>[KEY]>
+	}
+} =>
+	Object.fromEntries(
+		Object.entries(schema.shape).map(([key, shape]) => {
+			/* eslint-disable no-console */
+			if (DEBUG_MAKE_PROPS) console.log(`makeProp: generating “${key}”`)
+
+			const zodTypeSet = walkShape(shape)
+
+			/**
+			 * WORKAROUND: Usually, one should probably call shape.isOptional()
+			 *
+			 * However, this currently doesn’t seem to work with z.function().default()
+			 * and this also internally just calls safeParse on the object — evaluating
+			 * the default function, which wastes performance AND could cause
+			 * unintended side-effects
+			 */
+			const isOptional =
+				zodTypeSet.has('ZodDefault') || zodTypeSet.has('ZodOptional')
+
+			const prop: PropOptions<unknown, boolean> = {
+				validator: propValidator(shape, key),
+			}
+
+			if (isOptional) prop.default = shape._def.defaultValue
```

Comment:
check if zod adds the `undefined` by default to the `default` attribute
```
const i = z.string().optional() 
```
and i don't use a default 

i expect
```i._def.defaultValue.toBe(undefined)```

Reply:
check in unit tests @carsoli